compiler/qsc_passes/src/baseprofck.rs

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#[cfg(test)]
+mod tests;
+
+use miette::Diagnostic;
+use qsc_data_structures::span::Span;
+use qsc_hir::{
+    hir::{BinOp, CallableKind, Expr, ExprKind, Item, ItemKind, Lit, Package, SpecBody, SpecGen},
+    ty::{Prim, Ty},
+    visit::{walk_expr, walk_item, Visitor},
+};
+use thiserror::Error;
+
+#[derive(Clone, Debug, Diagnostic, Error)]
+pub enum Error {
+    #[error("cannot compare measurement results")]
+    #[diagnostic(help(
+        "comparing measurement results is not supported when performing base profile QIR generation"
+    ))]
+    #[diagnostic(code("Qsc.BaseProfCk.ResultComparison"))]
+    ResultComparison(#[label] Span),
+
+    #[error("result literals are not supported")]
+    #[diagnostic(help(
+        "result literals `One` and `Zero` are not supported when performing base profile QIR generation"
+    ))]
+    #[diagnostic(code("Qsc.BaseProfCk.ResultLiteral"))]
+    ResultLiteral(#[label] Span),
+
+    #[error("non-Result return type in entry expression")]
+    #[diagnostic(help(
+        "returning types other than Result from the entry expression is not supported when performing base profile QIR generation"
+    ))]
+    #[diagnostic(code("Qsc.BaseProfCk.ReturnNonResult"))]
+    ReturnNonResult(#[label] Span),
+
+    #[error("intrinsic operations that return types other than Result or Unit are not supported")]
+    #[diagnostic(help(
+        "intrinsic operations that return values other than Result or Unit are not supported when performing base profile QIR generation"
+    ))]
```

Comment:
You likely meant a text consistent between the lines 39 and 41:
```suggestion
    #[error("intrinsic operations that return types other than Result or Unit are not supported")]
    #[diagnostic(help(
        "intrinsic operations that return types other than Result or Unit are not supported when performing base profile QIR generation"
    ))]
```